Make tokenize CLI tool have nicer command line arguments.

Noeda opened the pull request:
Tokenize CLI tool is one of the tools in `examples/*`. It's a pretty short and simple tool that takes arguments like this:

```
  tokenize MODEL_FILENAME PROMPT [--ids]
```

And it would load the model, read the prompt, and then print a list of tokens it interpreted, or if --ids was given, just the integer values.

This changeset makes the command a bit more sophisticated with more options:

```
mikkojuola@Mikkos-Mac-Studio ~/llama.cpp> ./build/bin/tokenize
usage: ./build/bin/tokenize [options]

The tokenize program tokenizes a prompt using a given model,
and prints the resulting tokens to standard output.

It needs a model file, a prompt, and optionally other flags
to control the behavior of the tokenizer.

Invoke './build/bin/tokenize' like this:

    ./build/bin/tokenize MODEL_FNAME PROMPT [--ids]

  or this:

    ./build/bin/tokenize [options], where options are:

    -h, --help                           print this help and exit
    -m MODEL_PATH, --model MODEL_PATH    path to model.
    --ids                                if given, only print numerical token IDs, and not token strings.
    -f PROMPT_FNAME, --file PROMPT_FNAME read prompt from a file.
    -p PROMPT, --prompt PROMPT           read prompt from the argument.
    --stdin                              read prompt from standard input.
    --no-bos                             do not ever add a BOS token to the prompt, even if normally the model uses a BOS token.
    --log-disable
```

It will still recognize the old form (i.e. simple positional arguments) just to not surprise people. Although I would myself like to remove it entirely, to simplify the thing. Not sure anyone actually uses this tool except for ad-hoc testing like I do. Opinions on completely removing the "old style arguments"?

Motivation: I've been using this tool for my own tests with tokenization divergence investigations. I find it useful to do quick ad-hoc tests on text tokenization and comparisons. In particular I wanted it to behave nice if you give it a filename or pipe into it from `stdin`.

I took my hacks and cleaned them up into nicer looking command line arguments, following the style and argument names of some other CLI tools I saw. Also in general I added some error checking etc. so you are more likely to get a readable error than a segfault if you did something wrong.

Draft because I need to test some of the argument combinations and also Windows, and I want to see the CI results on GitHub here. I think the stdin reading as it is written might be sketchy on Windows, if you try to physically type letters, which would now become a feature of `tokenize`.

(`std::cin` does not have `.is_open()`? Got really confused when I was trying to write code to check did we read from `stdin` properly without syscall failures and trying to figure out if the code is checking syscall failures in a waterproof way. I'm a C programmer not a C++ one dammit)

Noeda committed:
Make tokenizer.cpp CLI tool nicer.

Before this commit, tokenize was a simple CLI tool like this:

  tokenize MODEL_FILENAME PROMPT [--ids]

This simple tool loads the model, takes the prompt, and shows the tokens
llama.cpp is interpreting.

This changeset makes the tokenize more sophisticated, and more useful
for debugging and troubleshooting:

  tokenize [-m, --model MODEL_FILENAME]
           [--ids]
           [--stdin]
           [--prompt]
           [-f, --file]
           [--no-bos]
           [--log-disable]

It also behaves nicer on Windows now, interpreting and rendering Unicode
from command line arguments and pipes no matter what code page the user
has set on their terminal.

Noeda commented:
Just noticed the CI doesn't run...maybe because it's my first PR and I'm not on an allowlist? Do I have a way to run the compilation tests somehow myself?

ggerganov commented on examples/tokenize/tokenize.cpp:
```diff
@@ -3,33 +3,259 @@
 #include <string>
 #include <vector>
 
+static void print_usage_information(const char* argv0, FILE* stream) {
```
Put spaces front and back - see the rest of the codebase:

```suggestion
static void print_usage_information(const char * argv0, FILE * stream) {
```

Noeda replied:
Changed all the `something*`s into `something *`s in the new push 👍

Noeda commented:
Added bunch of stuff since last commit, as part of wrestling with Windows cmd.exe console crap:

- A lot of new code to interpret and render characters properly on Windows. Good lord Windows is annoying. But you get correctly interpreted and rendered text now out-of-box without setting any code pages (although you may have to set a font depending on what text you use).
- `--ids` now prints in a format that parses directly as Python or JSON (useful for sketchy pipe shenanigans)
- `--log-disable` to silence `stderr` (consistent with `main`)
- Fixed the style, like the `*` pointer stuff to be more consistent with the rest of the codebase.
- Prints "failed utf-8 decode" on tokens that don't parse as UTF-8. Seems like a fairly common thing to happen with modern models where individual tokens don't decode to valid UTF-8. I made it print hex codes instead so you see the bytes it wants to decode as, even if we can't render them properly.

I noticed midway that we had similar code handling Windows stuff in `common/console.cpp`. Not exactly what I needed for the `tokenize` but I added a TODO comment about it, and made the Windows bits a bit more general so maybe a later contribution has an easier time moving that to common code.

---

## Some checking that all is good

Verified that tokens are interpreted the same way on `cmd.exe` and Linux, and also that Windows renders tokens correctly (when valid UTF-8):

Windows `cmd.exe`, with `--prompt こんにちは`

<img width="917" alt="Screenshot 2024-03-26 at 2 17 39 PM" src="https://github.com/ggerganov/llama.cpp/assets/833719/5408b385-4c5b-426d-9afb-45e5403435cb">

Reading from a file looks fine:

<img width="925" alt="Screenshot 2024-03-26 at 2 19 47 PM" src="https://github.com/ggerganov/llama.cpp/assets/833719/1393f8de-4187-4ad7-9204-3a268fe75561">

Piping on Windows works too:

<img width="924" alt="Screenshot 2024-03-26 at 2 20 11 PM" src="https://github.com/ggerganov/llama.cpp/assets/833719/d708a356-7e50-407b-b0f6-ce0dea92d916">

Checked that we get same IDs for `こんにちは` on Mac (got same on Linux too):

<img width="928" alt="Screenshot 2024-03-26 at 2 20 55 PM" src="https://github.com/ggerganov/llama.cpp/assets/833719/09a7251a-90f6-4588-b0df-1a3012775ad9">

---

If the CI doesn't complain and there's no other feedback to fix I'm done with the PR.

cebtenzzre commented on examples/tokenize/tokenize.cpp:
```diff
@@ -3,42 +3,439 @@
+        HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
+        DWORD dwMode = 0;
+
+        // According to Microsoft docs:
+        // "WriteConsole fails if it is used with a standard handle that is redirected to a file."
+        // Also according to the docs, you can use GetConsoleMode to check for that.
+        if (hConsole == INVALID_HANDLE_VALUE || !GetConsoleMode(hConsole, &dwMode)) {
+            printf("%s", str);
+            return;
+        }
```
Falling back to WriteFileA just because you can't WriteConsoleW isn't ideal. Why not use the standard wprintf() so it works regardless of redirection?

Noeda replied:
When I was testing this, it correctly piped non-ASCII text through the program on Windows. I think that condition (console reading failing) should only happen if you are piping. But not entirely sure; I don't typically develop on Windows so I could totally be wrong here.

I'm not that familiar with Windows but I got the sense that simple `printf` will just encoding-agnostically let bytes through in a pipe scenario where you are not involving the console. Do you know if that's correct? I actually looked at `WriteFile` first, but I couldn't tell if it was really any different from just doing a simpler `printf`, and thought a simple `printf` is much simpler. I didn't see if `WriteFileA` or `WriteFileW` even exist so I assumed `WriteFile` does not care about encodings, and is purely byte-oriented.

All the complicated Windows stuff is either about 1) handling command line argument encoding 2) rendering text correctly when it comes out on other side. (this function is 2)

Noeda replied:
Actually...I just realized I did not test the output pipe. Only input pipe, i.e. stdin. Okay, so that might definitely be wrong if printf doesn't work propely.

I'll come back to test this. If it doesn't work, I'll replace it.

cebtenzzre replied:
Sorry, there is no WriteFileA, but I wouldn't be surprised if printf() to a stream open in text mode is going to do a conversion from the system code page (e.g. Windows-1252) to UTF-16 internally.

Noeda replied:
Just tested it; I _think_ it's doing the same thing `WriteFile` would in this situation, i.e. it seems to behave (almost) encoding-agnostic with pipes. It does seem to turn `\n` into `\r\n`. Messing with `chcp` does not seem to affect results, got same md5sum for all outputs and it looks the same.

<img width="866" alt="command_example2" src="https://github.com/ggerganov/llama.cpp/assets/833719/0da67a86-4813-4738-912d-4241e81fc2e5">

When doing `cat` on Mac they look identical; but the Windows one has `\r\n` instead of just `\n` so it's following Windows style line endings (not visible in the screenshot below but you can see it in `hexedit` or have `vim` tell you). (out.txt = made with `tokenize` on mac, `out_windows.txt` = the one from above screenshots moved to Mac).

<img width="677" alt="Screenshot 2024-03-26 at 3 22 56 PM" src="https://github.com/ggerganov/llama.cpp/assets/833719/11be36d9-70f6-4a97-ae73-ab225eaf2e9a">

I checked with an extra `fprintf(stderr, ...)` in the code branch to check it's actually running the `printf`.

---

> Why not use the standard wprintf() so it works regardless of redirection?

When I was Googling around how to put Unicode stuff visible in `cmd.exe` no matter the code page reliably, I got the sense that `WriteConsoleW` and `wprintf` are not exactly the same, the former being less likely to do anything weird, since it works directly with the console itself. And you'd have to use `_setmode` with `wprintf`. But I admit didn't test the the wprintf approach, just went with the first implementation that worked which was the `WriteConsoleW` one.

cebtenzzre commented on examples/tokenize/tokenize.cpp:
```diff
@@ -3,42 +3,439 @@
+
+        // MultiByteToWideChar reports an error if str is empty, don't report
+        // them as invalid_utf8.
+        if (strlen(str) == 0) {
```
This would be more idiomatic:
```suggestion
        if (*str == '\0') {
```
(or `*str == 0`, or `!*str`, or `str[0] == '\0'`... you get the idea)

Noeda replied:
Fixed :+1:

Noeda committed:
style fix: strlen(str) == 0 --> *str == 0

ggerganov reviewed (approved):
> It will still recognize the old form (i.e. simple positional arguments) just to not surprise people. Although I would myself like to remove it entirely, to simplify the thing. Not sure anyone actually uses this tool except for ad-hoc testing like I do. Opinions on completely removing the "old style arguments"?

Yes, let's remove the old style arguments to simplify

Noeda committed:
Simplify tokenize.cpp; by getting rid of handling positional style arguments.

It must now be invoked with long --model, --prompt etc. arguments only.
Shortens the code.

ggerganov committed:
Merge branch 'master' into tokenizer-nicer-args

ggerganov commented on examples/tokenize/tokenize.cpp:
```diff
@@ -3,40 +3,391 @@
 
 #include <cmath>
 #include <cstdio>
+#include <iostream>
```
`<iostream>` should not be needed

mofosyne replied:
removed and then ran

cmake -B build
cmake --build build --config Release

and confirmed no compile issue.

mofosyne committed:
tokenize.cpp: iostream header no longer required

mofosyne commented:
Appears everyone is happy so far with this PR and will merge once CI issue in main branch is all sorted.
